crates/walrus-service/src/node.rs

```diff
@@ -2928,29 +3183,78 @@ impl ServiceState for StorageNodeInner {
         blob_id: BlobId,
         sliver_pair_index: SliverPairIndex,
         sliver: Sliver,
+        intent: UploadIntent,
     ) -> Result<bool, StoreSliverError> {
         self.check_index(sliver_pair_index)?;
+        let (metadata, persisted) = self
+            .resolve_metadata_for_sliver(&blob_id, intent.is_pending())
+            .await?;
 
-        ensure!(
-            self.is_blob_registered(&blob_id)?,
-            StoreSliverError::NotCurrentlyRegistered,
-        );
-
-        // Get metadata first to check encoding type.
-        let metadata = self
-            .storage
-            .get_metadata(&blob_id)
-            .context("database error when storing sliver")?
-            .ok_or(StoreSliverError::MissingMetadata)?;
-
-        // Check if encoding type is supported
         let encoding_type = metadata.metadata().encoding_type();
         if !encoding_type.is_supported() {
             return Err(StoreSliverError::UnsupportedEncodingType(encoding_type));
         }
 
-        self.store_sliver_unchecked(metadata, sliver_pair_index, sliver)
+        if persisted {
+            return self
+                .store_sliver_unchecked(metadata.as_ref().clone(), sliver_pair_index, sliver)
+                .await;
+        }
+
+        let shard_storage = self
+            .get_shard_for_sliver_pair(sliver_pair_index, &blob_id)
+            .await?;
+
+        let shard_status = shard_storage
+            .status()
             .await
+            .context("Unable to retrieve shard status")?;
+
+        if !shard_status.is_owned_by_node() {
+            return Err(
+                ShardNotAssigned(shard_storage.id(), self.current_committee_epoch()).into(),
+            );
+        }
+
+        if shard_storage
+            .is_sliver_type_stored(&blob_id, sliver.r#type())
+            .context("database error when checking sliver existence")?
+        {
+            return Ok(false);
+        }
+
+        let metadata_for_verification = metadata.as_ref().clone();
+        let (_, verified_sliver) = self
+            .verify_sliver_against_metadata(metadata_for_verification, sliver)
+            .await?;
+
```

Comment:
suggestion: this part is duplicated from `store_sliver_unchecked`. Can we extract them into a common function?

Reply:
done, moved to `prepare_sliver_for_storage`.